pkg/urbit/jets/e/en_json.c

```diff
@@ -0,0 +1,425 @@
+}
+
+static void
+_serialize_string(json_buffer *buf_u, u3_noun a)
+{
+  if ( _(u3du(a)) ) {
+    u3m_bail(c3__exit);
+  }
+
+  c3_w len_w = u3r_met(3, a);
+
+  _append_char(buf_u, '"');
+  for (c3_w i = 0; i < len_w; ++i) {
+    c3_y c_y = u3r_byte(i, a);
+
+    switch ( c_y ) {
+      case 0 ... 9:
+      case 11 ... 31: {
+        _append_text(buf_u, _JSON_UNICODES[c_y], 6);
+      } break;
+
+      case 10: {
+        _append_text(buf_u, _JSON_NEWLINE, sizeof(_JSON_NEWLINE) - 1);
+      } break;
+
+      case 34: {
+        _append_text(buf_u, _JSON_DOQ, sizeof(_JSON_DOQ) - 1);
+      } break;
+
+      case 92: {
+        _append_text(buf_u, _JSON_BAS, sizeof(_JSON_BAS) - 1);
+      } break;
+
+      case 127: {
+        _append_text(buf_u, _JSON_DEL, sizeof(_JSON_DEL) - 1);
+      } break;
+
+      default: {
+        _append_char(buf_u, c_y);
+      } break;
+    }
+  }
+  _append_char(buf_u, '"');
+}
+
+static void
+_serialize_array(json_buffer *buf_u, u3_noun a)
+{
+  _append_char(buf_u, '[');
+
+  if ( u3_nul != a ) {
+    u3_noun i, t = a;
+
+    while ( u3_nul != t ) {
+      u3x_cell(t, &i, &t);
+      _serialize(buf_u, i);
+      _append_char(buf_u, ',');
+    }
+    
+    // Remove trailing comma from array contents
+    --buf_u->len_w;
+  }
+
+  _append_char(buf_u, ']');
+}
+
+static void
+_serialize_object_helper(json_buffer *buf_u, u3_noun a)
+{
+  if ( u3_nul != a ) {
+    u3_noun n_a, l_a, r_a;
+    u3_noun pn_a, qn_a;
+    u3x_trel(a, &n_a, &l_a, &r_a);
+    u3x_cell(n_a, &pn_a, &qn_a);
+
+    // order is important to match unjetted tree traversal
+    //
+    _serialize_object_helper(buf_u, r_a);
+
+    _serialize_string(buf_u, pn_a);
+    _append_char(buf_u, ':');
+    _serialize(buf_u, qn_a);
+    _append_char(buf_u, ',');
+
+    _serialize_object_helper(buf_u, l_a);
+  }
+}
+
+static void
+_serialize_object(json_buffer *buf_u, u3_noun a)
+{
+  _append_char(buf_u, '{');
+
+  if ( u3_nul != a ) {
+    _serialize_object_helper(buf_u, a);
+
+    // Remove trailing comma from object contents
+    --buf_u->len_w;
+  }
+
+  _append_char(buf_u, '}');
+}
+
+static void
+_serialize(json_buffer *buf_u, u3_noun a)
+{
+  if ( u3_nul == a ) {
+      _append_text(buf_u, _JSON_NULL, sizeof(_JSON_NULL) - 1);
+  }
+  else {
+    u3_noun s, p;
+    u3x_cell(a, &s, &p);
+
+    switch ( s ) {
+      default: u3m_bail(c3__fail);
+
+      case c3__a: return _serialize_array(buf_u, p);
+      case c3__o: return _serialize_object(buf_u, p);
+      case c3__b: return _serialize_loobean(buf_u, p);
+      case c3__n: return _serialize_number(buf_u, p);
+      case c3__s: return _serialize_string(buf_u, p);
+    }
+  }
+}
+
+/* jet interface functions
+*/
+
+u3_atom
+u3qe_en_json(u3_noun a)
+{
+  c3_w siz_w = 0;
+  _measure(&siz_w, a);
+
+  json_buffer  bof_u;
+  json_buffer *buf_u = &bof_u;
+  c3_y buf_y[siz_w];
+
+  buf_u->buf_y = buf_y;
+  buf_u->len_w = 0;
+
+  _serialize(buf_u, a);
+
+  u3_atom res = u3i_bytes(buf_u->len_w, buf_u->buf_y);
```

Comment:
You can avoid allocating twice (as you've done here), and also skip the VLA by using the `slab` api:

```
u3i_slab sab_u;
u3i_slab_init(&sab_u, 3, siz_w);

// fill the slab by writing to sab_u.buf_y

return u3i_slab_mint_bytes(&sab_u);
```

And if your measurement is precise, you can skip the final trimming step with `u3i_slab_moot_bytes()`.

Reply:
Done